File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf-8').toString().trim().split('\n');
let idx = 0;
let answer = '';

while (input[idx] !== '.') {
    const str = input[idx];
    const stack = [];
    let flag = 'yes';
    
    for (let s of str) {
        if (s === '(' || s === '[') {
            stack.push(s);
        } else if (s === ')') {
            if (stack.pop() !== '(') {
                flag = 'no';
                break;
            }
        } else if (s === ']') {
            if (stack.pop() !== '[') {
                flag = 'no';
                break;
            }
        }
    }
    
    if (stack.length > 0) flag = 'no';
    
    answer += flag + '\n';
    idx++;
}

console.log(answer);

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

for line in sys.stdin.read().splitlines():
    if line == '.':
        break

    stack = []
    to_x = {'(':-1, ')':1, '[':-2, ']':2}
    flag = True
    
    for s in line:
        if s not in to_x:
            continue
            
        x = to_x[s]
        
        if x < 0:
            stack.append(x)
        else:
            if stack:
                y = stack[-1]
                
                if x + y != 0:
                    flag = False
                    break
                else:
                    stack.pop()
            else:
                flag = False
                break
                    
    print('yes' if (not stack and flag) else 'no')